### PR TITLE
Cleanup of simple_imputer

### PR DESCRIPTION
**Pull request by eddiebergman**
Does a clean-up of `SimpleImputer` stemming from the weirdness of the `if` statements, mentioned in #345. Does not change functionality.

* Simplifies the `fit` method a bit
* Add docstrings
* Fixes some typing
* Document unexpected functionality of passing `'constant'` for `categorical_strategy` instead of `'constant_!missing!'`.
    * `'constant'` - fill_value = 0  *(sklearn default)*
    * `'constant_!missing!'` - fill_value = -1

### CODE REVIEW

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 1]
```diff
@@ -15,92 +13,154 @@
...
-                 numerical_strategy: str = 'mean',
-                 categorical_strategy: str = 'most_frequent'):
+    def __init__(
+        self,
+        random_state: Optional[Union[np.random.RandomState, int]] = None,
+        numerical_strategy: str = 'mean',
+        categorical_strategy: str = 'most_frequent'
+    ):
+        """
+        Parameters
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:32]: Actually, we follow google style docstring. You can find more information about it [here](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). Also, you can check out the docstrings as they are used in AutoPyTorch [here](https://github.com/automl/Auto-PyTorch/blob/ef8d21ae982883548a8e77a3690eefa810d3ae07/autoPyTorch/api/base_task.py#L107).

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 2]
```diff
@@ -29,45 +29,36 @@ def __init__(
...
 
-        y: Optional[np.ndarray]
-            The labels for the input features `X`
-
-        Returns
-        -------
-        SimpleImputer
-            returns self
+        Returns:
+            SimpleImputer: returns the object itself
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:61]: Can you add a new line after `SimpleImputer:`? Otherwise, the formatting of documentation generated by Sphinx causes issues.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 3]
```diff
@@ -156,10 +144,8 @@ def get_properties(
     ) -> Dict[str, Union[str, bool]]:
         """Get the properties of the SimpleImputer class and what it can handle
 
-        Returns
-        -------
-        Dict[str, Union[str, bool]]
-            A dict from property names to values
+        Returns:
+            Dict[str, Union[str, bool]]: A dict from property names to values
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:148]: Same here

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 4]
```diff
@@ -15,92 +13,142 @@
...
+        that categorical data is fit with an Ordinal Scaler.
     """
 
-    def __init__(self,
-                 random_state: Optional[Union[np.random.RandomState, int]] = None,
-                 numerical_strategy: str = 'mean',
-                 categorical_strategy: str = 'most_frequent'):
+    def __init__(
+        self,
+        random_state: Optional[Union[np.random.RandomState, int]] = None,
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:27]: We removed integer type from `random_state`, so it must be `Optional[np.random.RandomState]`

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Hello thanks for the PR.
Since we had some changes (probably those are not in ASK), so I will leave some comments.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 5]
```diff
@@ -15,92 +13,142 @@
...
         dataset_properties: Optional[Dict[str, BaseDatasetPropertiesType]] = None,
-        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(hyperparameter='numerical_strategy',
-                                                                                  value_range=("mean", "median",
-                                                                                               "most_frequent",
-                                                                                               "constant_zero"),
-                                                                                  default_value="mean",
-                                                                                  ),
+        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter='numerical_strategy',
+            value_range=("mean", "median", "most_frequent", "constant_zero"),
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: It would be better if we can use `Enum`.
We did not have it so much just because I did not join the refactoring phase.
At least, I would like to reduce so many dependencies on hard-coded strings as much as possible.
Because this file also uses so many hard-coded strings, which we can avoid by enum.

```
from enum import Enum
from typing import List


class NumericalImputerChoices(Enum):
    mean = "mean"
    median = "median"
    most_frequent = "most_frequent"
    constant_zero = "constant_zero"

    @classmethod
    def get_choices(cls) -> List[str]:
        return [c.value for c in cls]


class CategoricalImputerChoices(Enum):
    most_frequent = "most_frequent"
    constant_missing = "constant_!missing!"

    @classmethod
    def get_choices(cls) -> List[str]:
        return [c.value for c in cls]
```

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 6]
```diff
@@ -15,92 +13,142 @@
...
+    def __init__(
+        self,
+        random_state: Optional[Union[np.random.RandomState, int]] = None,
+        numerical_strategy: str = 'mean',
+        categorical_strategy: str = 'most_frequent'
+    ):
+        """
+        Note:
+            Using 'constant' defaults to fill_value of 0 where 'constant_!missing!'
+            uses a fill_value of -1. This behaviour should probably be fixed.
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:34]: Is this what you mean?
```suggestion
            'constant' uses fill_value of 0 by default while
            'constant_!missing!' uses fill_value of -1. This behaviour should probably be fixed.
```

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 7]
```diff
@@ -15,92 +13,142 @@
...
+
+        Args:
+            random_state (Optional[Union[np.random.RandomState, int]]):
+                The random state to use for the imputer.
+            numerical_strategy (str: default='mean'):
+                The strategy to use for imputing numerical columns.
+                Can be one of ['mean', 'median', 'most_frequent', 'constant', 'constant_!missing!']
+            categorical_strategy (str: default='most_frequent')
+                The strategy to use for imputing categorical columns.
+                Can be one of ['mean', 'median', 'most_frequent', 'constant_zero']
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:44]: We can move this doc-string to the enums once we make them. 

Btw, choices of `categorical_strategy` is incompatible with the hyperparamer search space.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py [hunk 8]
```diff
@@ -15,92 +13,142 @@
...
+                `dataset_properties`
+        """
         cs = ConfigurationSpace()
-        assert dataset_properties is not None, "To create hyperparameter search space" \
-                                               ", dataset_properties should not be None"
-        if len(dataset_properties['numerical_columns']) \
-                if isinstance(dataset_properties['numerical_columns'], List) else 0 != 0:
+
+        if dataset_properties is None:
+            raise ValueError("SimpleImputer requires `dataset_properties` for generating"
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:125]: Could you add a test here?

**Review by nabenabe0928 — COMMENTED** [review]
Hi thanks for cleaning up this file.
It is much better than the previous state:)
I left some comments as well.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 4 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:27]: Okay, will do. As an fyi, this can cause different output if you run the same function twice on the same object. The random state produces a sequence of numbers.

For example, if you create a single RandomState object and pass it to every object that requires a random_state, you will get different output depending on the order in which objects use that random_state. On the flip-side, if you use an int, they are independant of each other and so it doesn't matter which order objects use it.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 5 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: I can do that, you do however require to update these whenever sklearn updates them as well as that is where they are forwarded to.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 7 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:44]: Sure, I'm just not sure what you mean by the second point, should I do something with respect to this?

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 8 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:125]: Sure, it wasn't tested before but I can add one.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 5 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: I can see why you want to do that but it makes the code that uses it a little less pretty. For example, it's hard to see where the error with the follow code is:

```python
        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(
            hyperparameter='numerical_strategy',
            value_range=NumericalImputerChoice.get_choices(),
            default_value=NumericalImputerChoice.mean,
        ),
```

It should be `default_value=NumericalImputerChoice.mean.value`. The problem here is that Enum's are namespaces `NumericalImputerChoice.mean == "NumericalImputerChoice.mean"` where as we would like `NumericalImputerChoice.mean == "mean"`. This is more readily achievable with a `NamedTuple`. I will implement the Enum version and you can change it if you like or leave it as is

Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: Another clean'ish solution is just ditch the enum part. 

```python
class Choices:
    x: str = "x"
    y: str = "y"
    
assert Choices.x = "x"
```

The type is still a string, meaning it's easy to use, people don't need to know about the existence of this class to use it, they can still pass a string. It also allows for the internal code to use the class, where we do know about it.

Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: As another note, it also makes parameters extremely long, it's 101 characters which is over the character count limit of the checker:
```python
    def __init__(
        self,
        random_state: Optional[np.random.RandomState] = None,
        numerical_strategy: NumericalImputerChoice = NumericalImputerChoice.mean.value,
        categorical_strategy: CategoricalImputerChoice = CategoricalImputerChoice.most_frequent.value
    ):
```

Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:97]: I've not changed it for now, there's too many decisions to make that I think can be addressed based on how you would like it done yourself. I've changed it back to strings.

Comment (ravinkohli) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:98]: Hey, thanks for the insights. Yeah, I think such changes should be a part of a separate PR as this PR is meant to clean up the messy statements in this file.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 4 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py]: Yeah, that is also true. My explanation was not sufficient, but we, in reality, decided to use `seed` for `int` and `random_state` for `np.random.RandomState`.
So it is a very good decision if we switch to `seed` instead of `random_state` in the future.

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 7 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py]: we have the following as the default searching space, right?
So I think either this value_range or your doc-string should follow the other.
I am assuming that you take the doc-string from AutoSKLearn, so probably this `value_range` should be replaced, don't you think?
```
        categorical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(
            hyperparameter='categorical_strategy',
            value_range=("most_frequent", "constant_!missing!"),
            default_value="most_frequent"
        )
```

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 4 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py]: We kept `random_state` as the arg name and allow for both, internally we just pass the `seed` argument given at construction of an `AutoSklearnClassifier` so that internally it's an int passed throughout.

We follow sklearn in principle so we copy [their expected behaviour](https://scikit-learn.org/stable/common_pitfalls.html#using-none-or-randomstate-instances-and-repeated-calls-to-fit-and-split).

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 5 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:98]: All the problems come from the fact that we still need to follow python3.7.
Once we switched to python3.8, we can use `Literal`...

Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:98]: We have the same thing, there is `Literal` at the moment but the type checking for it doesn't work as expected.

```python
def f(s: Literal['yes', 'no'])
   ...
   
f("yes") # Type Error

param: Literal = "yes"
f(param) # Okay
```

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 7 above)
Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py]: Ahh my bad, I see your point. I've updated the docstring rather than update the search space, just to be inline with what's there so the actual functionality doesn't change.

I actually looked at the `sklearn` docs for possible values

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py (hunk 5 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:98]: All the problems come from the fact that we still need to follow python3.7.
Once we switched to python3.8, we can use `Literal`...

> in general, enum values should never be directly used, such as their string value

Yeah, I agree with you. I am just uncomfortable using software that assumes that users google string choices (typically sklearn).
And I found it really useful to be able to use `enum` or `class` such as in numpy, facebook ax..
But I got your point, let's stay for now.
I did not know some points you mentioned, thanks for raising them. @eddiebergman .

(just a question, but) the better solution will be something like this:
```
class NumericalImputerChoices(NamedTuple):
    mean = "mean"
    median = "median"

num_imputer_choices = NumericalImputerChoices()
```
and use `str` for all the typings.

Btw, you do not need to put `.value` (FYI).

Comment (eddiebergman) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py:98]: also, you can add `python` after the backticks ``` of a code block to get python highlighting ;)

**Review by nabenabe0928 — APPROVED** [review]
